signer/xpi/cose.go

```diff
@@ -0,0 +1,339 @@
+		v = cose.ES384
+	case cose.ES512.Name:
+		v = cose.ES512
+	default:
+		v = nil
+	}
+	return v
+}
+
+// generateIssuerEEKeyPair returns a public and private key pair for
+// the provided COSEAlgorithm
+func (s *XPISigner) generateCOSEKeyPair(coseAlg *cose.Algorithm) (eeKey crypto.PrivateKey, eePublicKey crypto.PublicKey, err error) {
+	var signer *cose.Signer
+
+	switch coseAlg {
+	case nil:
+		err = fmt.Errorf("Cannot generate private key for nil cose Algorithm")
+	case cose.PS256:
+		const size = 2048
+		eeKey, err = s.getRsaKey(size)
+		if err != nil {
+			err = errors.Wrapf(err, "failed to generate rsa private key of size %d", size)
+			return
+		}
+		eePublicKey = eeKey.(*rsa.PrivateKey).Public()
+	case cose.ES256:
+		fallthrough
+	case cose.ES384:
+		fallthrough
+	case cose.ES512:
+		signer, err = cose.NewSigner(coseAlg, nil)
+		if err != nil {
+			err = errors.Wrapf(err, "failed to generate private key")
+			return
+		}
+		eeKey = signer.PrivateKey
+		eePublicKey = eeKey.(*ecdsa.PrivateKey).Public()
+	}
+	return
+}
+
+// isSupportedCOSEAlgValue returns whether the COSE alg value is supported or not
+func isSupportedCOSEAlgValue(algValue interface{}) bool {
+	return (algValue == cose.PS256.Value ||
+		algValue == cose.ES256.Value ||
+		algValue == cose.ES384.Value ||
+		algValue == cose.ES512.Value)
+}
+
+
+func expectHeadersAndGetKeyID(actual, expected *cose.Headers) (kidValue interface{}, err error) {
+	if actual == nil || expected == nil {
+		err = errors.New("xpi: cannot compare nil COSE headers")
+		return
+	}
+	if len(actual.Unprotected) != len(expected.Unprotected) {
+		err = fmt.Errorf("xpi: unexpected non-empty Unprotected headers got: %v", actual.Unprotected)
+		return
+	}
+	if len(actual.Protected) != len(expected.Protected) {
+		err = fmt.Errorf("xpi: unexpected Protected headers got: %v expected: %v", actual.Protected, expected.Protected)
+		return
+	}
+	if _, ok := expected.Protected[algHeaderValue]; ok {
+		algValue, ok := actual.Protected[algHeaderValue]
+		if !ok {
+			err = fmt.Errorf("xpi: missing expected alg in Protected Headers")
+			return
+		}
+		if !isSupportedCOSEAlgValue(algValue) {
+			err = fmt.Errorf("xpi: alg %v is not supported", algValue)
+			return
+		}
+	}
+	if _, ok := expected.Protected[kidHeaderValue]; ok {
+		kidValue, ok = actual.Protected[kidHeaderValue]
+		if !ok {
+			err = fmt.Errorf("xpi: missing expected kid in Protected Headers")
+			return
+		}
+	}
+	return
+}
+
+var (
+	expectedMessageHeaders = &cose.Headers{
+		Unprotected: map[interface{}]interface{}{},
+		Protected: map[interface{}]interface{}{
+			kidHeaderValue: nil,
+		},
+	}
+	expectedSignatureHeaders = &cose.Headers{
+		Unprotected: map[interface{}]interface{}{},
+		Protected: map[interface{}]interface{}{
+			kidHeaderValue: nil,
+			algHeaderValue: nil,
+		},
+	}
+)
+
+// isValidCOSESignature checks whether a COSE signature is valid for XPIs
+func isValidCOSESignature(sig *cose.Signature) (eeCert *x509.Certificate, err error) {
```

Comment:
I was expecting this function to actually verify a signature, but it looks like it's only checking some parameters and returning the EE cert. I'd suggest renaming it.

Reply:
to me validate != verify so these just check structure and not contents, and I need to add COSE verification using go-cose to the parent `verifyCOSESignatures` (since we need the cose manifest bytes and I didn't want to pass them everywhere).